docs/proposals/security-context-constraints.md

```diff
@@ -0,0 +1,320 @@
+that is exposed to end users (builds, deployments), the isolated end user namespaces in the cluster, and
+the individual users inside those namespaces.  Infrastructure components that operate on behalf of a
+user (builds, deployments) should be allowed to run at an elevated level of permissions without
+granting the user themselves an elevated set of permissions.
+
+## Goals
+
+1.  Associate [service accounts](http://docs.k8s.io/design/service_accounts.md) and users with
+a set of constraints that dictate how a [security context](./security_contexts.md) is established for a pod.
+1.  Provide the ability for users and infrastructure components to run pods with elevated privileges
+on behalf of another user or within a namespace where privileges are more restrictive.
+1.  Secure the ability to reference elevated permissions or to change the constraints under which
+a user runs.
+
+## Use Cases
+
+Use case 1:
+As an administrator, I can create a namespace for a person that can't create privileged containers
+AND enforces that the UID of the containers is set to a certain value
+
+Use case 2:
+As a cluster operator, an infrastructure component should be able to create a pod with elevated
+privileges in a namespace where regular users cannot create pods with these privileges or execute
+commands in that pod.
+
+Use case 3:
+As a cluster administrator, I can allow a given namespace (or service account) to create privileged
+pods or to run root pods
+
+Use case 4:
+As a cluster administrator, I can allow a project administrator to control the security contexts of
+pods and service accounts within a project
+
+
+## Requirements
+
+1.  Provide a set of restrictions that controlls how a security context is created as a new, cluster-scoped, object
+called SecurityContextConstraints.
+1.  User information in `user.Info` must be available to admission controllers. (Completed in
+https://github.com/GoogleCloudPlatform/kubernetes/pull/8203)
+1.  Some authorizers may restrict a user’s ability to reference a service account.  Systems requiring
+the ability to secure service accounts on a user level must be able to add a policy that enables
+referencing specific service accounts themselves.
+1.  Admission control must validate the creation of Pods against the allowed set of constraints.
+
+## Design
+
+### Model
+
+The model for a security context constraint is a combination of the security context constraint
+itself and cluster wide strategies enforced by an allocator.  The separation of concerns falls into
+two distinct categories: permissions that the pod can request and runtime parameters the cluster
+admin controls.
+
+An example of a permission that can be requested is to run a pod as privileged.  This is set on a
+pod by pod basis.  A cluster wide setting may be the allocation of user ids that pods must run under.
+The cluster administrator may decide that no pod can run as the root user in their cluster and
```

Comment:
The "how do I get a uid" strategy is indeed on the SCC as you suspect.  It is a strategy rather than a boolean to allow things like 

1.  must run as X uid that may be pre-allocated and assigned to a namespace
1.  may run as any non-root uid (this has kubelet implications when considering a USER directive may be in a docker file rather than specifically set on the request)
1.  must run as random uid in range

where the strategy is responsible for finding an allocated uid, generating one, etc and also validating that a request falls within the strategies acceptable boundaries.